### PR TITLE
[split 3/4] Refactor absorbed MLA projection handling

### PR DESCRIPTION
**Pull request by HollowMan6**
Split out from #3674 to reduce review scope.

Original changes by @HollowMan6 in #3674.

## Scope
- Switch absorbed MLA to a combined KV up-projection layout.
- Add helper handling for packed THD output rank restoration and optional core-attention V up-projection consumption.
- Pass CP communication metadata into absorbed MLA core attention.
- Add focused absorbed MLA tests for checkpoint metadata and updated projection-gradient parity.

## Dependencies
None.

## Validation
- `git diff --check upstream/main..split-3674-absorbed-mla`

### CODE REVIEW

File: megatron/core/transformer/experimental_attention_variant/absorbed_mla.py [hunk 1]
```diff
@@ -302,34 +352,19 @@ def __init__(
...
-            tp_group=pg_collection.tp,
-            name=(name + ".linear_k_up_proj") if name is not None else None,
-        )
-        self.linear_v_up_proj = build_module(
-            submodules.linear_v_up_proj,
+        self.linear_kv_up_proj = build_module(
+            submodules.linear_kv_up_proj,
             self.config.kv_lora_rank,
-            self.config.num_attention_heads * self.config.v_head_dim,
+            self.config.num_attention_heads * (self.config.qk_head_dim + self.config.v_head_dim),
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply megatron/core/transformer/experimental_attention_variant/absorbed_mla.py:358]: Fixed in `49ab87e53`. I restored a compatibility `_load_from_state_dict` path that combines old `linear_k_up_proj` / `linear_v_up_proj` weights into the new per-head `linear_kv_up_proj` layout before delegating to the base loader, and added focused coverage for that migration.

File: megatron/core/transformer/experimental_attention_variant/absorbed_mla.py [hunk 2]
```diff
@@ -643,41 +668,100 @@ def qkv_up_proj_and_rope_apply(q_compressed, kv_compressed, k_pos_emb, rotary_po
...
+            up_v_weight = inputs[5]
+            attn_mask_type = inputs[6]
             attn_mask_type = AttnMaskType(attn_mask_type.item())
             output_ = self.core_attention(
                 q_absorbed,
                 k_compressed,
-                v_compressed,
-                hidden_states,
-                q_compressed,
+                None,
```
Comment (santhnm2) [inline megatron/core/transformer/experimental_attention_variant/absorbed_mla.py]: Can you pass this with a keyword arg? (and the following `attention_mask`)

File: tests/unit_tests/transformer/experimental_attention_variant/test_absorbed_mla.py [hunk 3]
```diff
@@ -227,6 +230,96 @@ def get_mla_submodules(
...
+        packed_seq_params=packed_seq_params,
+    )
+
+    assert checkpoint_args is not None
+    assert all(arg is not packed_seq_params for arg in checkpoint_args)
+    assert all(arg is not None for arg in checkpoint_args)
+    assert output is hidden_states
+
+
+def test_load_from_state_dict_combines_split_kv_up_projection(monkeypatch):
```
Comment (santhnm2) [inline tests/unit_tests/transformer/experimental_attention_variant/test_absorbed_mla.py]: To confirm, this test assures that checkpoint load will remain backwards compatible?

**Review by santhnm2 — APPROVED** [review]

File: megatron/core/transformer/experimental_attention_variant/absorbed_mla.py (hunk 2 above)
Comment (HollowMan6) [reply megatron/core/transformer/experimental_attention_variant/absorbed_mla.py]: done

File: tests/unit_tests/transformer/experimental_attention_variant/test_absorbed_mla.py (hunk 3 above)
Comment (HollowMan6) [reply tests/unit_tests/transformer/experimental_attention_variant/test_absorbed_mla.py]: Yes, exactly. I just also renamed the test/docstring to make the backward-compatibility intent explicit.

**Review by Phlip79 — APPROVED** [review]

**Review by jaredcasper — APPROVED** [review]
